util/src/main/java/io/camunda/zeebe/util/startup/StartupProcess.java

```diff
@@ -0,0 +1,230 @@
+      shutdownFuture = new CompletableFuture<>();
+
+      proceedWithShutdown(context, shutdownFuture, new ArrayList<>());
+    } else {
+      logger.info("Shutdown already in progress");
+    }
+    return shutdownFuture;
+  }
+
+  private void proceedWithShutdown(
+      final CONTEXT context,
+      final CompletableFuture<CONTEXT> shutdownFuture,
+      final List<ShutdownExceptionRecord> collectedExceptions) {
+    if (startedSteps.isEmpty()) {
+      completeShutdownFuture(context, shutdownFuture, collectedExceptions);
+    } else {
+      final var stepToShutdown = startedSteps.pop();
+
+      logCurrentStep("Shutdown", stepToShutdown);
+      stepToShutdown
+          .shutdown(context)
+          .whenComplete(
+              (contextReturnedByShutdown, error) -> {
+                final CONTEXT contextToUse;
+                if (error != null) {
+                  collectedExceptions.add(new ShutdownExceptionRecord(stepToShutdown, error));
+                  contextToUse = context;
+                } else {
+                  contextToUse = contextReturnedByShutdown;
+                }
+
+                proceedWithShutdown(contextToUse, shutdownFuture, collectedExceptions);
+              });
+    }
+  }
+
+  private void completeShutdownFuture(
+      final CONTEXT context,
+      final CompletableFuture<CONTEXT> shutdownFuture,
+      final List<ShutdownExceptionRecord> collectedExceptions) {
+    if (collectedExceptions.isEmpty()) {
+      shutdownFuture.complete(context);
+      logger.info("Finished shutdown process");
+    } else {
+      final Throwable exception;
+      if (collectedExceptions.size() == 1) {
+        exception = collectedExceptions.get(0).getException();
+      } else {
+        final var details =
+            collectedExceptions.stream()
+                .map(entry -> entry.getStep().getName() + " " + entry.getException().getMessage())
+                .collect(Collectors.joining("\n"));
+        exception =
+            new Exception(
+                "Exceptions occurred during shutdown: \n"
+                    + details
+                    + "\nSee suppressed exceptions for more detail.");
+        collectedExceptions.stream()
+            .map(ShutdownExceptionRecord::getException)
+            .forEach(exception::addSuppressed);
+      }
+      shutdownFuture.completeExceptionally(exception);
+      logger.warn("Finished shutdown process with exception:", exception);
+    }
+  }
+
+  private void logCurrentStep(final String process, final StartupStep<CONTEXT> step) {
+    logger.info(process + " " + step.getName());
+  }
+
+  private static final class ShutdownExceptionRecord {
+    private final StartupStep<?> step;
+    private final Throwable exception;
+
+    private ShutdownExceptionRecord(final StartupStep<?> step, final Throwable exception) {
+      this.step = step;
+      this.exception = exception;
+    }
+
+    private StartupStep<?> getStep() {
```

Comment:
:upside_down_face: As it is used only in the same class and it's really just a tuple, I think we could just access the properties directly instead of via getters. But that's just a nit.